readthedocs/config/config.py

```diff
@@ -135,7 +121,108 @@ def catch_validation_error(self, key):
                 code=error.code,
                 error_message=str(error),
                 source_file=self.source_file,
-                source_position=self.source_position)
+                source_position=self.source_position
+            )
+
+    def validate(self):
+        raise NotImplementedError()
+
+    @property
+    def base(self):
+        raise NotImplementedError()
+
+    def output_base(self):
+        # TODO: this isn't used
+        raise NotImplementedError()
+
+    @property
+    def name(self):
+        raise NotImplementedError()
+
+    @property
+    def python(self):
+        raise NotImplementedError()
+
+    @property
+    def version(self):
+        raise NotImplementedError()
+
+    @property
+    def type(self):
+        raise NotImplementedError()
+
+    @property
+    def pip_install(self):
+        raise NotImplementedError()
+
+    @property
+    def install_project(self):
+        raise NotImplementedError()
+
+    @property
+    def extra_requirements(self):
+        raise NotImplementedError()
+
+    @property
+    def python_interpreter(self):
+        raise NotImplementedError()
+
+    @property
+    def python_version(self):
+        raise NotImplementedError()
+
+    @property
+    def python_full_version(self):
+        raise NotImplementedError()
+
+    @property
+    def use_system_site_packages(self):
+        raise NotImplementedError()
+
+    @property
+    def use_conda(self):
+        raise NotImplementedError()
+
+    @property
+    def conda_file(self):
+        raise NotImplementedError()
+
+    @property
+    def requirements_file(self):
+        raise NotImplementedError()
+
+    @property
+    def formats(self):
+        """Documentation formats to be built."""
+        raise NotImplementedError()
+
+    @property
+    def build_image(self):
+        raise NotImplementedError()
+
+
+class BuildConfig(BuildConfigBase):
+
+    """Version 1 of the configuration file."""
+
+    BASE_INVALID_MESSAGE = 'Invalid value for base: {base}'
+    BASE_NOT_A_DIR_MESSAGE = '"base" is not a directory: {base}'
+    NAME_REQUIRED_MESSAGE = 'Missing key "name"'
+    NAME_INVALID_MESSAGE = (
+        'Invalid name "{name}". Valid values must match {name_re}')
+    TYPE_REQUIRED_MESSAGE = 'Missing key "type"'
+    CONF_FILE_REQUIRED_MESSAGE = 'Missing key "conf_file"'
+    PYTHON_INVALID_MESSAGE = '"python" section must be a mapping.'
+    PYTHON_EXTRA_REQUIREMENTS_INVALID_MESSAGE = (
+        '"python.extra_requirements" section must be a list.')
+
+    PYTHON_SUPPORTED_VERSIONS = [2, 2.7, 3, 3.5]
+    DOCKER_SUPPORTED_VERSIONS = ['1.0', '2.0', 'latest']
+
+    def __init__(self, *args, **kwargs):
+        self._config = {}
+        super(BuildConfig, self).__init__(*args, **kwargs)
```

Comment:
It was in the beginning, but I moved it later. Since this is a _private_ field used to pass the data around, but I'm not sure if the others _config versions_ are going to use another pattern to pass the data around, so it makes sense to put it in the base class.